src/main/java/jnr/ffi/provider/jffi/BufferParameterStrategy.java

```diff
@@ -21,27 +21,94 @@
 import com.kenai.jffi.MemoryIO;
 import com.kenai.jffi.ObjectParameterType;
 
-import java.nio.Buffer;
+import java.nio.*;
 import java.util.EnumSet;
 
 /**
  *
  */
 public final class BufferParameterStrategy extends ParameterStrategy {
+    private static final int BYTE_POSITION_SHIFT = 0;
+    private static final int SHORT_POSITION_SHIFT = 1;
+    private static final int CHAR_POSITION_SHIFT = 1;
+    private static final int BOOLEAN_POSITION_SHIFT = 2;
+    private static final int INT_POSITION_SHIFT = 2;
+    private static final int FLOAT_POSITION_SHIFT = 2;
+    private static final int LONG_POSITION_SHIFT = 3;
+    private static final int DOUBLE_POSITION_SHIFT = 3;
+
     private final int shift;
 
     private BufferParameterStrategy(StrategyType type, ObjectParameterType.ComponentType componentType) {
         super(type, ObjectParameterType.create(ObjectParameterType.ObjectType.ARRAY, componentType));
         this.shift = calculateShift(componentType);
     }
 
-    public long address(Buffer buffer) {
-        return buffer != null && buffer.isDirect() ? MemoryIO.getInstance().getDirectBufferAddress(buffer) + (buffer.position() << shift) : 0L;
+    public static long address(ByteBuffer ptr) {
+        return address(ptr, BYTE_POSITION_SHIFT);
+    }
+
+    public static long address(ShortBuffer ptr) {
+        return address(ptr, SHORT_POSITION_SHIFT);
+    }
+
+    public static long address(CharBuffer ptr) {
+        return address(ptr, CHAR_POSITION_SHIFT);
+    }
+
+    public static long address(IntBuffer ptr) {
+        return address(ptr, INT_POSITION_SHIFT);
+    }
+
+    public static long address(FloatBuffer ptr) {
+        return address(ptr, FLOAT_POSITION_SHIFT);
+    }
+
+    public static long address(LongBuffer ptr) {
+        return address(ptr, LONG_POSITION_SHIFT);
+    }
+
+    public static long address(DoubleBuffer ptr) {
+        return address(ptr, DOUBLE_POSITION_SHIFT);
+    }
+
+    public static long address(Buffer buffer) {
+        if (buffer instanceof ByteBuffer) {
```

Comment:
Annoying that there's no way to get the size of the buffer's component type.

Reply:
Agreed. I put this overload in place as a fallback. I would expect native signatures to use a more specific Buffer subtype in most cases which avoids the overhead of this instanceof chain.